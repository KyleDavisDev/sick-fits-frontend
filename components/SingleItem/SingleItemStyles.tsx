import styled from "../../theme/theme";

export const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.boxShadow};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;
