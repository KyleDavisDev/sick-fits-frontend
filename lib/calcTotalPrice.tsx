export default function(items) {
  return items.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}
