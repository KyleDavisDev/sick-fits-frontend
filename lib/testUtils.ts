import casual from "casual";
import { ReactWrapper } from "enzyme";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  created: casual.unix_time,
  id: casual.uuid,
  price: 5000,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg"
});

const fakeUser = () => ({
  __typename: "User",
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: []
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words()
});

const fakeOrder = () => ({
  __typename: "Order",
  id: casual.uuid,
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2018-04 - 06T19: 24: 16.000Z",
  user: fakeUser()
});

const fakeCartItem = (overrides?: any) => ({
  __typename: "CartItem",
  id: casual.uuid,
  created: casual.unix_time,
  quantity: casual.integer(0, 15),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides
});

const fakeCart = (overrides?: any) => ({
  __typename: "Cart",
  id: casual.uuid,
  items: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
  user: fakeUser(),
  created: casual.unix_time,
  updated: casual.unix_time,
  isActive: true,
  ...overrides
});

// Fake LocalStorage
class LocalStorageMock {
  store: any;
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeCart,
  fakeOrder,
  fakeOrderItem,
  simulateInputChange
};
