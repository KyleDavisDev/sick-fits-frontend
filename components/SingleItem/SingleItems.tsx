import * as React from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import Head from "next/head";
import ErrorMessage from "../ErrorMessage/ErrorMessage";
import { SingleItemStyles } from "./SingleItemStyles";

export const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      largeImage
    }
  }
`;

export interface ISingleItemProps {
  id: string;
}

export interface ISingleItemState {}

export default class SingleItem extends React.Component<
  ISingleItemProps,
  ISingleItemState
> {
  constructor(props: ISingleItemProps) {
    super(props);

    this.state = {};
  }

  public render() {
    return (
      <Query query={SINGLE_ITEM_QUERY} variables={{ id: this.props.id }}>
        {payload => {
          if (payload.loading) return <p>Loading...</p>;
          if (payload.error) return <ErrorMessage error={payload.error} />;
          if (!payload.data.item)
            return <p>No Item found for ID: {this.props.id}</p>;
          const { item } = payload.data;
          return (
            <SingleItemStyles>
              <Head>
                <title>Sick Fits | {item.title}</title>
              </Head>
              <img src={item.largeImage} alt={item.title} />
              <div className="details">
                <h2>Viewing {item.title}</h2>
                <p>{item.description}</p>
              </div>
            </SingleItemStyles>
          );
        }}
      </Query>
    );
  }
}
