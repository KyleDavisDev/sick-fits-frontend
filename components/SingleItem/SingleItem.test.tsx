import "jsdom-global/register";
import * as React from "react";
import { mount } from "enzyme";
import wait from "waait";
import SingleItem, { SINGLE_ITEM_QUERY } from "./SingleItems";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeItem } from "../../lib/testUtils";

describe("<SingleItem />", () => {
  const item = fakeItem();
  const mocks = [
    {
      request: { query: SINGLE_ITEM_QUERY, variables: { id: item.id } },
      result: { data: { item } }
    }
  ];

  it("renders", () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id={item.id} />
      </MockedProvider>
    );
    expect(wrapper).toBeTruthy();
  });

  it("renders loading text", () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id={item.id} />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain("Loading...");
  });
});
