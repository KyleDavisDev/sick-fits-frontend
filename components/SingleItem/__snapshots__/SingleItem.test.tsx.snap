// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> renders matches snapshot 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;

exports[`<SingleItem /> renders matches snapshot 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`<SingleItem /> renders matches snapshot 3`] = `
<p>
  dogs
</p>
`;
